File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/users/UserDtos.kt
package app.pantopus.android.data.api.models.users

import app.pantopus.android.data.api.models.common.JsonValue
import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/**
 * Compact session-identity projection. Build via the companion helpers
 * from the richer login / profile responses.
 */
@JsonClass(generateAdapter = true)
data class UserDto(
    val id: String,
    val email: String,
    @Json(name = "display_name") val displayName: String?,
    @Json(name = "avatar_url") val avatarUrl: String?,
)

/**
 * Social-link bundle inside [UserProfile]. Route:
 * `backend/routes/users.js:1427`.
 */
@JsonClass(generateAdapter = true)
data class SocialLinks(
    val website: String?,
    val linkedin: String?,
    val twitter: String?,
    val instagram: String?,
    val facebook: String?,
)

/**
 * `GET /api/users/profile` user envelope — route
 * `backend/routes/users.js:1427`. Shape fields whose upstream type is
 * provider-dependent (residency, inviteProgress) are decoded as untyped.
 */
@JsonClass(generateAdapter = true)
data class UserProfile(
    val id: String,
    val email: String,
    val username: String,
    val firstName: String,
    val middleName: String?,
    val lastName: String,
    val name: String,
    val phoneNumber: String?,
    val dateOfBirth: String?,
    val address: String?,
    val city: String?,
    val state: String?,
    val zipcode: String?,
    val accountType: String,
    val role: String,
    val verified: Boolean,
    val residency: JsonValue?,
    @Json(name = "avatar_url") val avatarUrl: String?,
    @Json(name = "profile_picture_url") val profilePictureUrl: String?,
    val profilePicture: String?,
    val bio: String?,
    val tagline: String?,
    val socialLinks: SocialLinks?,
    val skills: List<String>?,
    @Json(name = "followers_count") val followersCount: Int?,
    @Json(name = "average_rating") val averageRating: Double?,
    @Json(name = "gigs_posted") val gigsPosted: Int?,
    @Json(name = "gigs_completed") val gigsCompleted: Int?,
    val profileVisibility: String?,
    val createdAt: String,
    val updatedAt: String,
)

/** Envelope for `GET /api/users/profile`. Route: `backend/routes/users.js:1427`. */
@JsonClass(generateAdapter = true)
data class ProfileResponse(
    val user: UserProfile,
    /** Shape varies by invite service — decoded lazily. */
    @Json(name = "invite_progress") val inviteProgress: JsonValue?,
)

/**
 * `PATCH /api/users/profile` request. Every field optional — unspecified
 * keys are left untouched server-side. Route:
 * `backend/routes/users.js:1503`.
 */
@JsonClass(generateAdapter = true)
data class ProfileUpdateRequest(
    val firstName: String? = null,
    val middleName: String? = null,
    val lastName: String? = null,
    val phoneNumber: String? = null,
    val address: String? = null,
    val city: String? = null,
    val state: String? = null,
    val zipcode: String? = null,
    val dateOfBirth: String? = null,
    val bio: String? = null,
    val tagline: String? = null,
    val profileVisibility: String? = null,
    val website: String? = null,
    val linkedin: String? = null,
    val twitter: String? = null,
    val instagram: String? = null,
    val facebook: String? = null,
)

/** Envelope for `PATCH /api/users/profile`. Route: `backend/routes/users.js:1503`. */
@JsonClass(generateAdapter = true)
data class ProfileUpdateResponse(
    val message: String,
    val user: UserProfile,
)

File: frontend/apps/ios/Pantopus/Core/Networking/Models/Users/UserDTOs.swift
//
//  UserDTOs.swift
//  Pantopus
//
//  DTOs for `GET /api/users/profile` and `PATCH /api/users/profile`.
//  Route handlers: `backend/routes/users.js:1427` and `:1503`.
//

import Foundation

/// Compact user identity used in app session state. Derived from the
/// authenticated-user payload; kept small so AuthManager doesn't leak the
/// full profile surface into every screen that only needs id/email.
public struct UserDTO: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    public let email: String
    public let displayName: String?
    public let avatarURL: URL?

    private enum CodingKeys: String, CodingKey {
        case id, email
        case displayName = "display_name"
        case avatarURL = "avatar_url"
    }

    public init(id: String, email: String, displayName: String?, avatarURL: URL?) {
        self.id = id
        self.email = email
        self.displayName = displayName
        self.avatarURL = avatarURL
    }

    /// Project a rich [`AuthenticatedUser`](x-source-tag://AuthenticatedUser) down to the session shape.
    public init(from authUser: AuthenticatedUser) {
        self.id = authUser.id
        self.email = authUser.email
        self.displayName = authUser.name.isEmpty ? nil : authUser.name
        self.avatarURL = nil
    }

    /// Project a full [`UserProfile`](x-source-tag://UserProfile) down to the session shape.
    public init(from profile: UserProfile) {
        self.id = profile.id
        self.email = profile.email
        self.displayName = profile.name
        if let raw = profile.avatarURL ?? profile.profilePictureURL, let url = URL(string: raw) {
            self.avatarURL = url
        } else {
            self.avatarURL = nil
        }
    }
}

/// Social-link bundle emitted in the profile response.
/// Route: `backend/routes/users.js:1427`.
public struct SocialLinks: Decodable, Sendable, Hashable {
    public let website: String?
    public let linkedin: String?
    public let twitter: String?
    public let instagram: String?
    public let facebook: String?
}

/// `GET /api/users/profile` user envelope — see `backend/routes/users.js:1427`.
///
/// Fields whose upstream shape is provider-dependent (e.g. `residency`,
/// `inviteProgress`) are modelled as `JSONValue?` rather than invented.
public struct UserProfile: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    public let email: String
    public let username: String
    public let firstName: String
    public let middleName: String?
    public let lastName: String
    public let name: String
    public let phoneNumber: String?
    public let dateOfBirth: String?
    public let address: String?
    public let city: String?
    public let state: String?
    public let zipcode: String?
    public let accountType: String
    public let role: String
    public let verified: Bool
    public let residency: JSONValue?
    public let avatarURL: String?
    public let profilePictureURL: String?
    public let profilePicture: String?
    public let bio: String?
    public let tagline: String?
    public let socialLinks: SocialLinks?
    public let skills: [String]?
    public let followersCount: Int?
    public let averageRating: Double?
    public let gigsPosted: Int?
    public let gigsCompleted: Int?
    public let profileVisibility: String?
    public let createdAt: String
    public let updatedAt: String

    private enum CodingKeys: String, CodingKey {
        case id, email, username, firstName, middleName, lastName, name
        case phoneNumber, dateOfBirth, address, city, state, zipcode
        case accountType, role, verified, residency
        case avatarURL = "avatar_url"
        case profilePictureURL = "profile_picture_url"
        case profilePicture
        case bio, tagline, socialLinks, skills
        case followersCount = "followers_count"
        case averageRating = "average_rating"
        case gigsPosted = "gigs_posted"
        case gigsCompleted = "gigs_completed"
        case profileVisibility, createdAt, updatedAt
    }
}

/// Envelope for `GET /api/users/profile` — route `backend/routes/users.js:1427`.
public struct ProfileResponse: Decodable, Sendable, Hashable {
    public let user: UserProfile
    /// Shape varies by invite service; decode lazily.
    public let inviteProgress: JSONValue?

    private enum CodingKeys: String, CodingKey {
        case user
        case inviteProgress = "invite_progress"
    }
}

/// `PATCH /api/users/profile` — see `backend/routes/users.js:1503`. Every
/// field is optional; unspecified keys are left untouched server-side.
public struct ProfileUpdateRequest: Encodable, Sendable, Hashable {
    public var firstName: String?
    public var middleName: String?
    public var lastName: String?
    public var phoneNumber: String?
    public var address: String?
    public var city: String?
    public var state: String?
    public var zipcode: String?
    public var dateOfBirth: String?
    public var bio: String?
    public var tagline: String?
    public var profileVisibility: String?
    public var website: String?
    public var linkedin: String?
    public var twitter: String?
    public var instagram: String?
    public var facebook: String?

    public init(
        firstName: String? = nil,
        middleName: String? = nil,
        lastName: String? = nil,
        phoneNumber: String? = nil,
        address: String? = nil,
        city: String? = nil,
        state: String? = nil,
        zipcode: String? = nil,
        dateOfBirth: String? = nil,
        bio: String? = nil,
        tagline: String? = nil,
        profileVisibility: String? = nil,
        website: String? = nil,
        linkedin: String? = nil,
        twitter: String? = nil,
        instagram: String? = nil,
        facebook: String? = nil
    ) {
        self.firstName = firstName
        self.middleName = middleName
        self.lastName = lastName
        self.phoneNumber = phoneNumber
        self.address = address
        self.city = city
        self.state = state
        self.zipcode = zipcode
        self.dateOfBirth = dateOfBirth
        self.bio = bio
        self.tagline = tagline
        self.profileVisibility = profileVisibility
        self.website = website
        self.linkedin = linkedin
        self.twitter = twitter
        self.instagram = instagram
        self.facebook = facebook
    }
}

/// Envelope for `PATCH /api/users/profile` — route `backend/routes/users.js:1503`.
public struct ProfileUpdateResponse: Decodable, Sendable, Hashable {
    public let message: String
    public let user: UserProfile
}
